File: reactant.ts
'use strict'

if(typeof require !== 'undefined')
    var EventDispatcher_:any = require('./eventdispatcher')
else
    var EventDispatcher_:any = EventDispatcher

class Reactant {
    unlink
    lastValue
    super_proc
    listen
    
    constructor(value?) {
        EventDispatcher_.call(this)
        this.value = value
    }
    
    assignment(value) {
        this.setFunc(() => value)
    }
    
    setFunc(func, onMod?) {
        var removal = this.setFuncSilent(func, onMod)
        this.proc()
        return removal
    }
    setFuncSilent(func, onMod?) {
        if(this.unlink)
            this.unlink()

        this.valueFunc = func

        // TODO: should only link when there are listeners on this reactant
        if(onMod)
            this.unlink = onMod.listen(() => this.proc())
        else
            this.unlink = undefined
        return this.unlink
    }
    depend(transformation, reactant) {
        this.setFunc((() => transformation(reactant.value)), reactant)
    }
    echo(reactant) {
        this.setFunc((() => reactant.value), reactant)
    }
    
    proc() {
        var prev = this.lastValue
        var curr = this.value
        if(this.equality(prev, curr))
            return

        // must be saved here because it might be altered during event propagation
        this.lastValue = curr
        
        this.super_proc(prev, curr)
    }
    
    listen_pc(callback) {
        var result = this.listen(callback)
        callback(undefined, this.value)
        // Don't have to check to see if the callback was removed because its removal wasn't made
        // accessible to any other stack frame yet.
        return result
    }
    and(other) {
        return this.compose(other, (a, b) => a && b)
    }
    or(other) {
        return this.compose(other, (a, b) => a || b)
    }
    transform(func) {
        var result = new Reactant()
        result.depend(func, this)
        return result
    }
    onCondition(predicate) {
        var result = new EventDispatcher_()
        this.listen((prev, curr) => {
            if(predicate(prev, curr))
                result.proc()
        })
        return result
    }
    on(target) {
        return this.onCondition((prev, curr) =>  this.equality(curr, target))
    }
    onNot(target) {
        return this.onCondition((prev, curr) => !this.equality(curr, target))
    }
    onCondition_pc(predicate, callback) {
        var result = this.onCondition(predicate).listen(callback)
        if(predicate(undefined, this.value))
            callback()
        return result
    }
    on_pc(target, callback) {
        return this.onCondition_pc(((prev, curr) => this.equality(curr, target)), callback)
    }
    compose(b, func) {
        var a = this
        var result = new Reactant()
        result.setFunc(() => func(a.value, b.value))

        ////////////////////////////////////////////////////// TODO: These listeners need to be removed when `result` has none of its own listeners, and re-added when it does
        var removeA = a.listen(result.proc.bind(result))
        var removeB = b.listen(result.proc.bind(result))

        return result
    }
    
    get value() {
        // Cannot return lastValue directly because reactants whose values depend on this
        // reactant cannot fire events correctly under the current implementation without
        // calling the valueFunc each time.
        return this.valueFunc()
    }
    set value(value) {
        this.assignment(value)
    }
    get not() {
        return this.transform(a => !a)
    }
    get negative() {
        return this.transform(a => -a)
    }
    get anyTruthy() {
        return this.onCondition((prev, curr) => curr)
    }
    get anyFalsy() {
        return this.onCondition((prev, curr) => !curr)
    }
    
    equality(a, b) {
        if(a === b)
            return true
        if(a && a.equals)
            return a.equals(b)
        if(b && b.equals)
            return b.equals(a)
        return false
    }
    valueFunc() {
        return undefined
    }
}
Reactant.prototype['__proto__'] = EventDispatcher_.prototype
Reactant.prototype.super_proc = EventDispatcher_.prototype.proc

if(typeof module !== 'undefined')
    module.exports = Reactant

File: src/esquire/Reactant.kt
package esquire


data class Change<T>(val prev: T, val curr: T)

class Reactant<T>(initialValue: T) {
    constructor(initialDefinition: ()->T,
                initialOnMod: EventDispatcher<Unit>? = null): this(initialDefinition()) {
        define(initialDefinition, initialOnMod)
    }

    private var definition: ()->T

    private var definitionReg: Registration? = null

    fun define(definition: ()->T, onMod: EventDispatcher<Unit>? = null) {
        this.definition = definition

        if(definitionReg != null)
            definitionReg!!.remove()

        if(onMod == null)
            definitionReg = null
        else
            definitionReg = onMod.listen { changed() }

        changed()
    }

    var value: T
        get() = definition()
        set(newValue) = define({ newValue })

    private var lastValue: T

    init {
        definition = { initialValue }
        lastValue = value

        js("""Kotlin.modules.esquire.esquire.Reactant.prototype.and =
              function(other) { return Kotlin.modules.esquire.esquire.and(this, other) };""")
        js("""Kotlin.modules.esquire.esquire.Reactant.prototype.or =
              function(other) { return Kotlin.modules.esquire.esquire.or(this, other) };""")
        js("""Kotlin.modules.esquire.esquire.Reactant.prototype.not =
              function() { return Kotlin.modules.esquire.esquire.not(this) };""")

        js("""Kotlin.modules.esquire.esquire.Reactant.prototype.minus =
              function(other) { return Kotlin.modules.esquire.esquire.minus_1(this, other) };""")
        js("""Kotlin.modules.esquire.esquire.Reactant.prototype.plus =
              function(other) { return Kotlin.modules.esquire.esquire.plus(this, other) };""")
        js("""Kotlin.modules.esquire.esquire.Reactant.prototype.div =
              function(other) { return Kotlin.modules.esquire.esquire.div(this, other) };""")
        js("""Kotlin.modules.esquire.esquire.Reactant.prototype.times =
              function(other) { return Kotlin.modules.esquire.esquire.times(this, other) };""")
        js("""Kotlin.modules.esquire.esquire.Reactant.prototype.negative =
              function() { return Kotlin.modules.esquire.esquire.minus(this) };""")
    }

    private var changes = EventDispatcher<Change<T>>()
    private var unitChanges = changes.toUnit()

    private fun changed() {
        val prev = lastValue
        val curr = value
        if(prev == curr)
            return

        // Must be saved here because it might be altered during event propagation
        lastValue = curr

        changes(Change(prev, curr))
    }

    fun listen(callback: (T, T)->Unit) = changes.listen { callback(it.prev, it.curr) }

    fun listenAndInvoke(callback: (T, T)->Unit): Registration {
        val curr = value
        callback(curr, curr)

        return listen(callback)
    }

    fun on(predicate: (T, T)->Boolean): EventDispatcher<Unit> {
        val result = EventDispatcher<Unit>()
        listen { prev, curr ->
            if(predicate(prev, curr))
                result(Unit)
        }
        return result
    }
    fun on(target: T) = on { prev, curr -> curr == target }
    fun onNot(target: T) = on { prev, curr -> curr != target }

    fun depend<U>(other: Reactant<U>, transformation: (U)->T) {
        define({ transformation(other.value) }, other.unitChanges)
    }
    fun echo(other: Reactant<T>) {
        define({ other.value }, other.unitChanges)
    }

    fun compose<U, V>(other: Reactant<U>, transformation: (T, U)->V): Reactant<V> {
        return Reactant<V>({ transformation(this.value, other.value) },
                unitChanges aggregate other.unitChanges)
    }

    fun transform<U>(transformation: (T)->U) = Reactant<U>({ transformation(value) }, unitChanges)

    // TODO
//    companion object {
//        fun tuple<A, B>(a: Reactant<A>, b: Reactant<B>): Reactant<Pair<A, B>> {
//            return Reactant({ Pair(a.value, b.value) }, a.unitChanges aggregate b.unitChanges)
//        }
//        fun tuple<A, B, C>(a: Reactant<A>, b: Reactant<B>, c: Reactant<C>): Reactant<Triple<A, B, C>> {
//            return Reactant({ Triple(a.value, b.value, c.value) },
//                            EventDispatcher.any(a.unitChanges, b.unitChanges, c.unitChanges))
//        }
//    }
}

fun Reactant<Boolean>.and(other: Reactant<Boolean>) = this.compose(other, { a, b -> a and b })
fun Reactant<Boolean>.or(other: Reactant<Boolean>) = this.compose(other, { a, b -> a or b })
fun Reactant<Boolean>.not() = this.transform({ !it })

fun Reactant<Double>.minus() = this.transform({ -it })
fun Reactant<Double>.minus(other: Reactant<Double>) = this.compose(other, { a, b -> a - b })
fun Reactant<Double>.plus(other: Reactant<Double>) = this.compose(other, { a, b -> a + b })
fun Reactant<Double>.times(other: Reactant<Double>) = this.compose(other, { a, b -> a * b })
fun Reactant<Double>.div(other: Reactant<Double>) = this.compose(other, { a, b -> a / b })
